lib/compilers/rust.ts

```diff
@@ -222,6 +222,16 @@ export class RustCompiler extends BaseCompiler {
         return options;
     }
 
+    override preProcess(source: string, filters: CompilerOutputOptions): [string, CompilerOutputOptions] {
+        // #4910, #5350: Avoid rustc warnings when building source with `fn main()` as `--crate-type lib`
+        const fnMainRe = /fn\s+main\(/;
+        if (source.match(fnMainRe)) {
+            filters.binary = true;
```

Comment:
Are you forcing the link when there's a main function in the input source? Or maybe I'm not reading this correctly?

Reply:
@dkm I re-read now the previous discussions, in particular https://github.com/compiler-explorer/compiler-explorer/issues/4910#issuecomment-1489164222, and I probably misunderstood it originally. The logic you guys agreed on was 'if `fn main()` is detected*, use `--crate-type bin` *AND* `--emit asm`' ? 

-* need to also exclude appearance in comments. The default rust code includes it.

Reply:
I'm not reading it like that. We can use `--crate-type bin` for the common case (nothing selected in the filters). As we get the assembly using `--emit asm`, we won't display external code, but we'll benefit from `--crate-type bin` not garbage collecting the `fn main`, if any. In particular, I'm not reading that we are trying to use different options based on the detection of `main` in the code. The only part where we look for `main` is when the user selects `link to binary` and we need to stub it with something if not present (and this is optional as it's not always as trivial as in C).

(sorry for the delay, the notification got flooded and I missed it)